internal/edittree/edittree.go

```diff
@@ -0,0 +1,1165 @@
+					rewritesComposites = append(rewritesComposites, i)
+					continue
+				}
+				panic(fmt.Errorf("invalid index %d during Insert operation", i))
+			}
+		}
+		// Do rewrites to clear out the newly-removed element.
+		e.deleteChildValue(idx)
+		for i := 0; i < len(rewritesScalars); i++ {
+			originalIdx := rewritesScalars[i]
+			rewriteIdx := rewritesScalars[i] - 1
+			v := e.childScalarValues[originalIdx]
+			e.deleteChildValue(originalIdx)
+			e.setChildScalarValue(rewriteIdx, v)
+		}
+		for i := 0; i < len(rewritesComposites); i++ {
+			originalIdx := rewritesComposites[i]
+			rewriteIdx := rewritesComposites[i] - 1
+			v := e.childCompositeValues[originalIdx]
+			e.deleteChildValue(originalIdx)
+			e.setChildCompositeValue(rewriteIdx, v)
+		}
+
+		// "bleed through" to the underlying array if needed.
+		// To do this, we sum up the zeroes below the current index, and use that value
+		// to index through the `eliminated` array until we find a surviving index.
+		if e.insertions.Element(idx) == 0 {
+			zeroesSeen := 1 + sumZeroesBelowIndex(idx, e.insertions)
+			// Mark appropriate position in `eliminated` array, or error.
+			elimIdx, found := findIndexOfNthZero(zeroesSeen, e.eliminated)
+			if !found {
+				panic(fmt.Errorf("could not successfully eliminate index %d from array", idx))
+			}
+			e.eliminated.Set(1, elimIdx)
+		}
+		// Delete element from insertions array, bump bit-vec over by 1.
+		e.insertions.Delete(idx)
+		return e, nil
+	default:
+		// Catch all primitive types.
+		return nil, fmt.Errorf("expected composite type, found value: %v (type: %T)", e.value.Value, e.value.Value)
+	}
+}
+
+//gcassert:inline
+func sumZeroesBelowIndex(index int, bv *bitvector.BitVector) int {
+	zeroesSeen := 0
+	for i := 0; i < index; i++ {
+		if bv.Element(i) == 0 {
+			zeroesSeen++
+		}
+	}
+	return zeroesSeen
+}
+
+func findIndexOfNthZero(n int, bv *bitvector.BitVector) (int, bool) {
+	zeroesSeen := 0
+	for i := 0; i < bv.Length(); i++ {
+		if bv.Element(i) == 0 {
+			zeroesSeen++
+		}
+		if zeroesSeen == n {
+			return i, true
+		}
+	}
+	return 0, false
+}
+
+// Helper function for sets/objects when the key isn't present in either
+// child map.
+func (e *EditTree) fallbackDelete(key *ast.Term) (*EditTree, error) {
+	value, err := e.value.Value.Find(ast.Ref{key})
+	if err != nil {
+		return nil, fmt.Errorf("cannot delete child key %v that does not exist", key)
+	}
+	keyHash, _ := e.getKeyHash(key)
+	e.setChildKey(keyHash, key)
+	if isComposite(ast.NewTerm(value)) {
+		e.setChildCompositeValue(keyHash, nil)
+	} else {
+		e.setChildScalarValue(keyHash, nil)
+	}
+	return NewEditTree(ast.NewTerm(value)), nil
+}
+
+// Unfurls a chain of EditTree nodes down a given path, or else returns an error.
+func (e *EditTree) Unfold(path ast.Ref) (*EditTree, error) {
+	switch {
+	// 0 path segments base case. (Root hits this.)
+	case len(path) == 0:
+		return e, nil
+	// 1+ path segment case.
+	case len(path) >= 1:
```

Comment:
🤔 Could we drop the case and unindent the rest? Looks like there's not much else going on here.

Reply:
Done.